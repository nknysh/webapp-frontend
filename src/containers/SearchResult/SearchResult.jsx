import React, { useState } from 'react';
import { compose, gt, head, last, map, partial, path, pathOr, prop, values, propOr } from 'ramda';
import { useTranslation } from 'react-i18next';
import { format } from 'date-fns';

import { ProductTypes } from 'config/enums';
import { getNumberOfDays, mapWithIndex, addDaysUTC } from 'utils';

import connect from './SearchResult.state';
import { propTypes, defaultProps } from './SearchResult.props';
import {
  StyledCard,
  CardImage,
  CardChipAvailability,
  CardChip,
  CardPrice,
  CardDetails,
  CardTitle,
  CardRating,
  CardPreferred,
  CardStarRating,
  CardStar,
  CardStarText,
  CardHighlights,
  CardHighlight,
  CardChipStack,
  ToolTip,
  PriceBreakdown,
  PriceBreakdownItem,
} from './SearchResult.styles';
import { formatPrice } from 'utils';
import LinkButton from 'components/LinkButton';
import { Link } from 'react-router-dom';

const getStartDateEndDate = dates => {
  const startDate = head(dates);
  const endDate = addDaysUTC(last(dates), 1);

  return { startDate, endDate };
};

const renderAmenity = (amenity, hotelName) => <CardHighlight key={`${hotelName}:${amenity}`}>{amenity}</CardHighlight>;

const renderRoomsBreakdown = (t, { title, dates = [], subProducts, ...product }) => {
  const { startDate, endDate } = getStartDateEndDate(dates);

  const count = getNumberOfDays({ startDate, endDate });
  const mealPlan = pathOr('', [ProductTypes.MEAL_PLAN, 0, 'product', 'meta', 'categoryType'], subProducts);

  return (
    <PriceBreakdownItem key={path(['product', 'uuid'], product) + startDate + endDate}>
      {t('labels.nightsIn', {
        count,
        nights: t('night', { count }),
        title,
        mealPlan,
        startDate: format(startDate, 'YYYY-MM-DD'),
        endDate: format(endDate, 'YYYY-MM-DD'),
      })}
    </PriceBreakdownItem>
  );
};

const renderTransfersBreakdown = (t, { title, dates = [], ...product }) => {
  const { startDate, endDate } = getStartDateEndDate(dates);

  return (
    <PriceBreakdownItem key={path(['product', 'uuid'], product) + startDate + endDate}>{title}</PriceBreakdownItem>
  );
};

const renderPrice = (t, { offerCount, response, showDiscountedPrice, currencyCode }) =>
  response && (
    <CardChipStack>
      <CardChip>
        {pathOr(true, ['totals', 'oneOrMoreItemsOnRequest'], response) ? (
          t('labels.onRequest')
        ) : (
          <ToolTip
            label={
              <CardPrice data-discounted={showDiscountedPrice && gt(offerCount, 0)}>
                {`${currencyCode}${
                  showDiscountedPrice
                    ? formatPrice(path(['totals', 'totalBeforeDiscount'], response))
                    : formatPrice(path(['totals', 'total'], response))
                }`}
              </CardPrice>
            }
          >
            <b>{t('labels.priceBasedOn')}</b>
            <PriceBreakdown>
              {map(
                partial(renderRoomsBreakdown, [t]),
                pathOr([], ['potentialBooking', ProductTypes.ACCOMMODATION], response)
              )}
              {map(
                partial(renderTransfersBreakdown, [t]),
                pathOr([], ['potentialBooking', ProductTypes.TRANSFER], response)
              )}
            </PriceBreakdown>
          </ToolTip>
        )}
      </CardChip>
      {showDiscountedPrice && gt(offerCount, 0) && (
        <CardChip>
          {pathOr(true, ['totals', 'oneOrMoreItemsOnRequest'], response) ? (
            t('labels.onRequest')
          ) : (
            <CardPrice data-discount={true}>{`${currencyCode}${formatPrice(
              path(['totals', 'total'], response)
            )}`}</CardPrice>
          )}
        </CardChip>
      )}
    </CardChipStack>
  );

const renderOfferBreakdown = (t, { offer }, i) => (
  <PriceBreakdownItem key={i + prop('uuid', offer)}>{prop('name', offer)}</PriceBreakdownItem>
);

const renderOffers = (t, { offerCount, offers }) => {
  return (
    gt(offerCount, 0) && (
      <CardChip data-secondary={true}>
        <ToolTip label={<span>{t('offerWithCount', { count: offerCount })}</span>}>
          {values(mapWithIndex(partial(renderOfferBreakdown, [t]), offers))}
        </ToolTip>
      </CardChip>
    )
  );
};

const renderAvailabilityChip = (t, { availableToHold }) => {
  return (
    availableToHold && (
      <CardChipAvailability>
        <ToolTip label={<span>{t('availability')}</span>}>{t('availableToHold')}</ToolTip>
      </CardChipAvailability>
    )
  );
};

export const SearchResult = ({
  bookingBuilder = {},
  currencyCode,
  featuredPhoto,
  name,
  offerCount,
  offers,
  preferred,
  showDiscountedPrice,
  starRating,
  id,
  highlights,
  overview,
}) => {
  const { t } = useTranslation();

  const { response } = bookingBuilder;

  const [isShowingHighlights, setIsShowingHighlights] = useState(false);

  const availableToHold = propOr(false, 'availableToHold', response);

  return (
    <StyledCard>
      <Link to={`/hotels/${id}`}>
        <CardImage style={{ backgroundImage: `url(${prop('url', featuredPhoto)})` }}>
          {preferred && <CardPreferred>{t('labels.preferred')}</CardPreferred>}
          {renderPrice(t, { response, offerCount, showDiscountedPrice, currencyCode })}
          {renderOffers(t, { offers, response, offerCount })}
          {renderAvailabilityChip(t, { availableToHold })}
        </CardImage>
      </Link>
      <CardDetails>
        <Link to={`/hotels/${id}`}>
          <CardTitle>{name}</CardTitle>
          <CardRating>
            <CardStarRating>
              <CardStar>star</CardStar>{' '}
              <CardStarText>
                {starRating} {t('star')}
              </CardStarText>
            </CardStarRating>
          </CardRating>
          <CardHighlights>{overview && overview.map(a => renderAmenity(a, name))}</CardHighlights>
        </Link>
        {!isShowingHighlights && (
          <LinkButton
            aria-label={t('labels.showHighlights')}
            data-role="linkButton.showHighlights"
            title={t('labels.showHighlights')}
            role="button"
            onClick={() => setIsShowingHighlights(true)}
          >
            + {t('labels.showHighlights')}
          </LinkButton>
        )}
        {isShowingHighlights && (
          <LinkButton
            aria-label={t('labels.hideHighlights')}
            data-role="linkButton.hideHighlights"
            title={t('labels.hideHighlights')}
            role="button"
            onClick={() => setIsShowingHighlights(false)}
          >
            - {t('labels.hideHighlights')}
          </LinkButton>
        )}
        {isShowingHighlights && (
          <CardHighlights>{highlights && highlights.map(a => renderAmenity(a, name))}</CardHighlights>
        )}
      </CardDetails>
    </StyledCard>
  );
};

SearchResult.propTypes = propTypes;
SearchResult.defaultProps = defaultProps;

export default compose(connect)(SearchResult);
