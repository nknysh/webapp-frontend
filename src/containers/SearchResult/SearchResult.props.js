import PropTypes from 'prop-types';

export const propTypes = {
  featuredPhoto: PropTypes.object,
  name: PropTypes.string,
  preferred: PropTypes.bool,
  promotionalText: PropTypes.string,
  starRating: PropTypes.string,
  suitableForHoneymooners: PropTypes.bool,
  listPrice: PropTypes.string,
  amenities: PropTypes.array,
  showDiscountedPrice: PropTypes.bool,
};

export const defaultProps = {
  featuredPhoto: {},
  name: '',
  preferred: false,
  promotionalText: '',
  suitableForHoneymooners: false,
  listPrice: '',
  additionalInfo: '',
  amenities: [],
  showDiscountedPrice: true,
};
