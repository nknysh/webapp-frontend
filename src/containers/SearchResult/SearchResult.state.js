import { connect } from 'react-redux';

import { getBookingAppliedOffers, getBookingAppliedOffersCount } from 'store/modules/bookings/selectors';

export const mapStateToProps = (state, { id }) => ({
  offers: getBookingAppliedOffers(state, id),
  offerCount: getBookingAppliedOffersCount(state, id),
});

export default connect(mapStateToProps);
