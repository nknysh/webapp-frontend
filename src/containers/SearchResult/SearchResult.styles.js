import styled from 'styled-components';
import { Icon } from '@material-ui/core';
import { Chip, ToolTip as BaseToolTip } from '@pure-escapes/webapp-ui-components';

import Card from 'components/Card';
import { theme } from 'styles';

export const StyledCard = styled(Card)`
  position: relative;
  margin: 0 0 ${theme.spacing.gutter * 2}px;
  padding-bottom: ${theme.spacing.gutter * 1.6}px;

  ${props => props.theme.breakpoints.tablet`
      margin ${theme.spacing.gutter}px;
    `}
`;

export const CardImage = styled(Card.Image)`
  position: relative;
  border-bottom: 1px solid ${theme.borders.default};
`;

export const CardChip = styled(Chip)``;

export const CardChipAvailability = styled(Chip)`
  color: ${props => props.theme.palette.available};
`;

export const CardPrice = styled(Card.Price)``;

export const CardDetails = styled.div`
  padding: ${theme.spacing.gutter * 2}px;
`;

export const CardTitle = styled(Card.Title)`
  margin-bottom: ${theme.spacing.gutter}px;
`;

export const CardDescription = styled(Card.Description)`
  border: 0;
  margin: 0;
`;

export const CardText = styled.div`
  color: ${theme.palette.neutral};
`;

export const CardPreferred = styled.div`
  background: ${theme.palette.primary};
  font-size: ${theme.fonts.sizes.normal}px;
  font-weight: ${theme.fonts.bold};
  position: absolute;
  top: 0;
  left: ${theme.spacing.gutter * 2}px;
  padding: ${theme.spacing.gutter}px ${theme.spacing.gutter * 2}px;
  color: ${theme.colors.white};
  text-transform: uppercase;
`;

export const CardCta = styled.div`
  padding: 0 ${theme.spacing.gutter * 2}px ${theme.spacing.gutter * 2}px;
`;

export const CardRating = styled.div`
  padding: 0;
  display: flex;
  align-items: center;
  border-bottom: 1px solid ${theme.borders.default};
  padding-bottom: ${theme.spacing.gutter}px;
  margin-bottom: ${theme.spacing.gutter}px;
`;

export const CardStarRating = styled.div`
  font-size: ${theme.fonts.sizes.normal}px;
  color: ${theme.palette.primary};
  font-weight: ${theme.fonts.bold};
  text-transform: uppercase;
  flex: 1;
  display: flex;
  align-items: center;
`;

export const CardStar = styled(Icon)`
  fill: ${theme.palette.primary};
  margin-right: ${theme.spacing.gutter / 2}px;
  font-size: ${theme.fonts.sizes.bigger}px;
`;

export const CardStarText = styled.span`
  flex: 1;
  font-size: ${theme.fonts.sizes.default}px;
  letter-spacing: 0.38px;
  line-height: ${theme.fonts.sizes.normal}px;
`;

export const Columns = styled.div`
  ${props => props.theme.breakpoints.tablet`
        display: flex;
    `}
`;

export const Column = styled.div`
  ${props => props.theme.breakpoints.tablet`
        flex: 1;
    `}
`;

export const CardSecondaryRating = styled.div`
  font-weight: ${theme.fonts.bold};
  text-transform: uppercase;
  color: ${theme.colors.marine};
  font-size: ${theme.fonts.sizes.default}px;
  letter-spacing: 0.38px;
  line-height: ${theme.fonts.sizes.normal}px;
  width: 50%;
`;

export const CardHighlights = styled.div`
  color: ${theme.colors.black};
  font-size: ${theme.fonts.sizes.normal}px;
  text-transform: uppercase;
  padding-top: ${theme.spacing.gutter}px;
  margin-bottom: ${theme.spacing.gutter}px;

  ${props => props.theme.breakpoints.tablet`
    display: flex;
    flex-wrap: wrap;
  `}
`;

export const CardHighlight = styled.div`
  margin: ${theme.spacing.gutter / 2}px 0;
  display: block;
  font-size: ${theme.fonts.sizes.default}px;

  ${props => props.theme.breakpoints.tablet`
    flex: 1 1 50%;
    width: 50%;
  `}
`;

export const CardAdditionalInfo = styled.div`
  display: block;
  padding-top: ${theme.spacing.gutter}px;
`;

export const CardAdditional = styled.p`
  display: block;
  margin: ${theme.spacing.gutter}px 0;
  font-size: ${theme.fonts.sizes.normal}px;
  color: ${theme.palette.neutral};

  &:last-child {
    margin: 0;
  }
`;

export const ToolTip = styled(BaseToolTip)`
  padding: 0;
`;

export const PriceBreakdown = styled.ul`
  margin: ${theme.spacing.gutter}px;
  padding: 0;
`;

export const PriceBreakdownItem = styled.li`
  margin-bottom: ${theme.spacing.gutter / 2}px;
`;

export const CardChipStack = styled.div`
  margin: 0 ${theme.spacing.gutter / 2}px;

  > div {
    margin: ${theme.spacing.gutter}px 0 0;
  }
`;
