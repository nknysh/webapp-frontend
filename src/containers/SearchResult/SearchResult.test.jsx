import React from 'react';

import { SearchResult } from './SearchResult';

const defaultProps = {
  hotel: {
    image: '',
    name: '',
    preferred: true,
    promotionalText: '',
    starRating: 5,
    suitableForHoneymooners: false,
    listPrice: '123',
    additionalInfo: [],
    features: [],
  },
};

const getComponent = props => shallow(<SearchResult {...defaultProps} {...props} />);

describe('<SearchResult />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
    });
  });
});
