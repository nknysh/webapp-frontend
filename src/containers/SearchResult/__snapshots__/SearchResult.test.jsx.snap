// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResult /> render matches snapshot 1`] = `
<SearchResultstyles__StyledCard>
  <Link
    replace={false}
    to="/hotels/undefined"
  >
    <SearchResultstyles__CardImage
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    />
  </Link>
  <SearchResultstyles__CardDetails>
    <Link
      replace={false}
      to="/hotels/undefined"
    >
      <SearchResultstyles__CardTitle />
      <SearchResultstyles__CardRating>
        <SearchResultstyles__CardStarRating>
          <SearchResultstyles__CardStar>
            star
          </SearchResultstyles__CardStar>
           
          <SearchResultstyles__CardStarText>
             
            Star
          </SearchResultstyles__CardStarText>
        </SearchResultstyles__CardStarRating>
      </SearchResultstyles__CardRating>
      <SearchResultstyles__CardHighlights />
    </Link>
    <LinkButton
      aria-label="Show Highlights"
      data-role="linkButton.showHighlights"
      onClick={[Function]}
      role="button"
      title="Show Highlights"
    >
      + 
      Show Highlights
    </LinkButton>
  </SearchResultstyles__CardDetails>
</SearchResultstyles__StyledCard>
`;
